### DISCUSSION TITLE
Bundled Library, NPM Package, Non-Module Version, Etc.

### DISCUSSION THREAD

**Discussion by SE2Dev** [biodiv/contactjs#24] [Ideas]
When migrating the library to TypeScript, I also added a bundler to our build configuration to allow bundling the source modules and type definitions into a much cleaner output. As per discussions [here](https://github.com/biodiv/contactjs/discussions/23), we may still want to provide a non-module version of the library as well.

My reasoning for choosing Parcel over something like Webpack mainly boiled down to the fact that in situations where there are name conflicts in bundled type declarations, Parcel automatically generates separate names for the types that share the same name. When trying to bundle type declarations with most other type bundlers right now, there are no checks performed to ensure that the bundled type declarations are actually valid; in the case of internal name conflicts, they wouldn't be valid.

After thinking about this some more, I've come to the conclusion that we don't really _need_ a bundler for the [distributed NPM package](https://www.npmjs.com/package/contactjs) since anyone using actually that package would most likely be using a bundler at the end of their build pipeline. Bundling ContactJS, itself, would be largely superfluous in cases such as this and actually make tree shaking somewhat more difficult. Publishing the built modules, and type declarations without bundling would still provide the same level of access for package users, as well as resolve the type declaration bundling issue mentioned above. For a non-module version of the library, however, bundling is still necessary - however it's somewhat reasonable to assume that anyone using the non-module version would not need access to the type declarations (since they'd require manual setup on their end).

Ultimately I think it may be in our best interest to switch to Webpack in order to provide a non-module version of the library under GitHub Releases, and provide un-bundled versions of the library via NPM. Does this sound reasonable?

**Comment by biodiv** [comment]
Thanks for elaborating, it sounds very reasonable in my opinion. It was the first time I used parcel and I liked it. However, if webpack can produce a non-module version, the switch will be worth it I think.

**Comment by GitMax42** [comment]
After a lot of research and testing, I found a **killer argument for using modules - dynamic import to load the module only when it is needed**
Dynamic module import is a function-like expression that allows loading an ECMAScript module asynchronously and dynamically into a potentially non-module environment.

The contact library makes only sense (for me!) on a touch-enabled device so I do not want to load it generally. My tested implementation is:

```
if(window.matchMedia("(pointer: coarse)").matches) {
   import("./contact.min.js")
   .then((contact) => {
      const element = document.querySelector("#s");
      const options = {
        supportedGestures : [contact.Pan, ],
        handleTouchEvents: false,
      };
      const l = new contact.PointerListener(element, options);
      element.addEventListener('swiperight', function() { doSwipeRight(); });
   })
   .catch((err) => {
      console.log("Error:" + err.message);
   });
} else {
   console.log("no touch device -> contact module not loaded!");
}
```


Remarks:
- the Script containing this dynamic loading does NOT need to be a module, and can be deferred
- this is ES6 standard and support is very good  [caniuse.com/es6-module-dynamic-import](https://caniuse.com/es6-module-dynamic-import)
- I still think it makes sense, to provide a non-module version for some implementations 

@SE2Dev  / @biodiv  : what do you think about this? Do you think the detection of touch devices via window.matchMedia is sufficient?

**Comment by biodiv** [comment]
Unfortunately, I have no experience with `MatchMedia` and `pointer:coarse`, but your approach looks good to me. It should be sufficient if you somehow catch `click` events instead of `tap` events when contact is not loaded. A `tap` is always a `click`, too.
